### ISSUE TITLE
[SIP-113] Calendar Heatmap Improvement 

### ISSUE THREAD

**Issue by Mattc1221** [apache/superset#26229]
## The problem

Calendar Heatmaps in Superset don't allow users to select the start and end of the heatmap range even though the calendar heatmap library that we use has this functionality. Documentation is [here](https://cal-heatmap.com/docs/options/scale#domain)

### Full Motivation

**<details><summary>Data with specific thresholds</summary>**
<p>

If your data has specific thresholds or boundaries that are meaningful for your analysis, manually setting the range allows you to highlight these thresholds accurately. For example, in a temperature heat map, you might want to distinguish between temperatures above and below freezing point with different colors.

</p>
</details> 

**<details><summary>Highlighting anomalies</summary>**
<p>

When you want to focus on anomalies or outliers in your data, manually setting the range can help make these anomalies more apparent. For instance, in a sales data heat map, you might want to highlight exceptionally high or low sales figures.

</p>
</details> 

**<details><summary>Comparing different time periods</summary>**
<p>

When comparing data across different time periods, setting a consistent range ensures that the color scale is uniform. This is especially important when comparing multiple calendar heat maps side by side.

</p>
</details> 

**<details><summary>Emphasizing specific values</summary>**
<p>

To emphasize specific values or a particular range of values that are particularly relevant to your analysis, manually setting the range allows you to create a visualization that aligns with your objectives.

</p>
</details> 

**<details><summary>Ensuring consistent color representation</summary>**
<p>

When working with multiple heat maps or sharing them with others, manually setting the range helps maintain consistency in color representations, making it easier for viewers to interpret the data.

</p>
</details> 

**<details><summary>Accommodating user preferences</summary>**
<p>

Depending on your audience or user preferences, manually setting the range can allow for customization. Users may have different color perception or need to see the data in a specific way, and manual range setting provides flexibility.

</p>
</details> 

**<details><summary>Sensitivity to small changes</summary>**
<p>

In certain scientific or engineering applications, you might want to create heat maps that are sensitive to even small changes in data values. Manually setting the range allows for fine-grained control to detect subtle variations.

</p>
</details> 

**<details><summary>Clarity and readability</summary>**
<p>

For educational or presentation purposes, manually setting the range can enhance the clarity and readability of the heat map, making it easier for the audience to understand and interpret the information.

</p>
</details> 

**<details><summary>Highlighting specific events or patterns</summary>**
<p>

Manually setting the range can be used to emphasize specific events or patterns in the data. For example, in a calendar heat map showing website traffic, you might want to highlight days with significant spikes in traffic.

</p>
</details> 

**<details><summary>Adherence to domain-specific standards</summary>**
<p>

In some domains, there may be established standards or conventions for color-coding data in heat maps. Manually setting the range allows you to adhere to these standards accurately.

</p>
</details> 

### TLDR

An example of why we would want to specify the calendar heatmap range:

1. I want to display 2 different calendar heatmaps that span 2 different time spans.
2. I want to see the differences between the two time spans.
3. With the current implemtation we would get something like this:
<img width="265" alt="Screenshot 2023-11-08 at 11 07 34 PM" src="https://github.com/apache/superset/assets/36670322/cb3defdd-64b3-45e7-99e4-e2f1ecd0199c">
<img width="233" alt="Screenshot 2023-11-08 at 11 08 35 PM" src="https://github.com/apache/superset/assets/36670322/0a20032c-3d38-4bf9-a409-a4a0b39f8213">
4. These 2 maps look similar and at first glance convey that they are displaying the same data. However, the color range on these 2 charts don't match and can thus easily lead the user to bad data analaysis

## Proposed Solution

1. Add a required chart control with the label `Metric Range`
2. The options for `Metric Range` would be `Automatic`, `Manual`
    - _Note: the default value would be `Automatic` and use the same strategy we use now_
3. Selecting `Manual` would display 2 new inputs: `Metric Range Start` and `Metric Range End` that would be integer inputs

https://github.com/apache/superset/assets/36670322/af22eda0-aab7-4b3d-8e98-57166966db68

**Comment by rusackas** [comment]
Thanks for the SIP (and the PR!). We're holding the PR while the SIP settles out. The next step here is to start a [DISCUSS] thread on the mailing list. Let me know if you want any assistance with that, you can just DM me on [slack](http://bit.ly/join-superset-slack).

**Comment by michael-s-molina** [comment]
> Alternatively, we could go forward with an ECharts-based calendar heatmap plugin. That might be smaller/lighter/maintainable, but it probably won't be able to reach feature parity with what you're proposing here.

I would vote for moving to the [ECharts plugin](https://echarts.apache.org/examples/en/editor.html?c=calendar-heatmap) unless there's a really good reason that blocks us from moving into that direction. I even have a [Heatmap PR](https://github.com/apache/superset/pull/25353) (not calendar) that will land soon and may help with the calendar version. We need to think about UI/UX consistency and code complexity and try to always move to ECharts to materialize [SIP-50](https://github.com/apache/superset/issues/10418).

**Comment by JohnDietrich-Pepper** [comment]
@rusackas @michael-s-molina 

Right now, the calendar heatmap in Superset has some severe limitations, particularly around consistency when multiple heatmaps are on the same dashboard that this SIP would address.  We can wait until the ECharts one is implemented but that could literally be a year or more away whereas this and the corresponding pull request could address that issue in months or less making the platform significantly more robust.

I realize you really want to move everything to ECharts but I personally think feature enhancements that allow Superset to be a viable alternative to other business intelligence platforms should come before trying to get everything on one Javascript library.  ECharts has some good features but it also has some areas that aren't as evolved as controls like the one calendar heatmap is based on.  I would vote for implementing the best library for a particular visualization.

**Comment by rusackas** [comment]
> I realize you really want to move everything to ECharts

Not at all the case... We prefer using it when it makes sense, but there are plenty of plugins where it might not be the best choice. 

> I would vote for implementing the best library for a particular visualization.

100% agreed.

The main worry in my book is extending the features of a legacy plugin that doesn't use the current data endpoint or provide other first-class features like drilling. I just hope that gets prioritized sometime before the legacy endpoint is deprecated.

**Comment by michael-s-molina** [comment]
> I would vote for implementing the best library for a particular visualization.

Any technical decision is always a trade-off. Any time you introduce a new library you need to consider all base logic that needs to be supported consistently throughout the app. Think about theming, tooltip handling, legend interaction, zooming, data transformation, etc. If you introduce a new library you need to implement these behaviors in a consistent manner with the rest of the plugins, to avoid decreasing UX. At times, you want to introduce a new library because it provides many important features that are not supported by the common library and the cost of implementing the base logic pays off. I generally use the principle where if you can provide the plugin in a satisfactory manner using ECharts, that should be the preferred way. Only when this is not an option, I would introduce a new library.

**Comment by JohnDietrich-Pepper** [comment]
What are the next steps on this?  We have dashboards deployed for customers that could leverage the work that Mattc1221 has already completed and would like to be able to use his updates without trying to fork and maintain concurrency with the official releases.

**Comment by rusackas** [comment]
Re-numbering this one to SIP-113, since another Issue was also numbered as 111, and went up for a DISCUSS thread on the official mailing list. Please proceed with opening a DISCUSS thread for this one when you want to move it forward.

**Comment by rusackas** [comment]
Discuss thread opened officially on the dev@ list :)

**Comment by rusackas** [comment]
@Mattc1221 I can carry this forward for a vote, if you're still hoping to implement it :)

**Comment by Mattc1221** [comment]
@rusackas that would be great! I would plan on just updating [this PR](https://github.com/apache/superset/pull/26230)

**Comment by rusackas** [comment]
Just got another piece of feedback that one of the major complaints with this chart is the orientation. It would be less confusing to users if there was a way to orient the calendar horizontally (like a wall calendar), or at least provide a control with the option for that sort of layout.

**Comment by rusackas** [comment]
I think this just needs one more binding vote to get passed. Please feel free to respond on the thread and call for help!

**Comment by rusackas** [comment]
Just pinged the thread on the Dev mailing list to try to get one more binding vote for this to pass. Anyone reading this (but especially PMC members) please feel free to vote!

**Comment by Mattc1221** [comment]
@rusackas how do you vote?

**Comment by rusackas** [comment]
Hi! To vote, you just reply to the Vote thread on the dev mailing list. If you're not on the list, you can join by emailing dev-subscribe@superset.apache.org. To see older messages, you can visit https://lists.apache.org/list.html?dev@superset.apache.org

**Comment by rusackas** [comment]
As it stands, there are 2 binding +1 votes (3 needed) and 2 non-binding +1 votes.  I'll try to rally some folks for a third.

**Comment by rusackas** [comment]
Got the third vote! Closing the thread, the vote has PASSED.

**Comment by rusackas** [comment]
This SIP has been inactive for quite a while. @Mattc1221 do you think you might circle back to this, or should we consider it "Discarded?"

**Comment by zakaryanla** [comment]
This improvement would be very useful for us too.
